sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala

```diff
@@ -81,11 +83,22 @@ case class ShuffledHashJoinExec(
 
   protected override def doExecute(): RDD[InternalRow] = {
     val numOutputRows = longMetric("numOutputRows")
+    val spillThreshold = getSpillThreshold
+    val inMemoryThreshold = getInMemoryThreshold
+    val streamSortPlan = getStreamSortPlan
+    val buildSortPlan = getBuildSortPlan
+    val fallbackSMJPlan = SortMergeJoinExec(leftKeys, rightKeys, joinType, condition, left, right)
+
     streamedPlan.execute().zipPartitions(buildPlan.execute()) { (streamIter, buildIter) =>
-      val hashed = buildHashedRelation(buildIter)
-      joinType match {
-        case FullOuter => fullOuterJoin(streamIter, hashed, numOutputRows)
-        case _ => join(streamIter, hashed, numOutputRows)
+      buildHashedRelation(buildIter) match {
+        case r: UnfinishedUnsafeHashedRelation =>
+          joinWithSortFallback(streamIter, buildIter, r.destructiveValues(), streamSortPlan,
```

Comment:
@c21 This fallback logic has been already deployed in your production? I just want to know that this can work well for real workloads (This can cause high performance penalties if the fallback happens only in a single task).

Reply:
@maropu - yes, more data is in https://github.com/apache/spark/pull/32210#issuecomment-823503243 .

> This can cause high performance penalties if the fallback happens only in a single task

Wondering why it is the case?

Reply:
For example, we have three tasks for a shuffle hash join. If the two tasks have no fallback and the last one has the fallback, the running time of the fallback one (building time of a hash table and sorting time of left/right input rows) can be much longer than the other two non-fallback ones? Or, I misunderstood the current logic?

Reply:
For runtime, yes. The total query run-time is dominated by the last finished task runtime. Just to point it out in case, without this change, this would be task and query failure.

Reply:
> @maropu - yes, more data is in #32210 (comment) .

Ah, thanks. I missed that comment.

Reply:
> For runtime, yes. The total query run-time is dominated by the last finished task runtime. Just to point it out in case, without this change, this would be task and query failure.

Yea, I basically agree to make the shuffle hash-join more robust (since a user possibly use inappropriate join hints in some cases). What I'm interested in is that; is there any other faster ballback logic than the current approach.

Reply:
> is there any other faster ballback logic than the current approach.

@maropu - yes, I am open and welcome any suggestion as well. Thanks for more insights from @sigmod and you @maropu . I will take a deeper look of hybrid join as in documentations you guys provided, and get back to you guys later.